sound/soc/sof/pm.c

```diff
@@ -138,25 +146,38 @@ static int sof_restore_pipelines(struct snd_sof_dev *sdev)
 		}
 	}
 
-	/* restore dai links */
+	link_array = devm_kzalloc(sdev->dev, sizeof(int) * (max_index + 1),
+				  GFP_KERNEL);
+	if (!link_array)
+		return -ENOMEM;
+
+	/* restore dai links from index 0 to max index */
 	list_for_each_entry_reverse(dai, &sdev->dai_list, list) {
 		struct sof_ipc_reply reply;
 		struct sof_ipc_dai_config *config = dai->dai_config;
 
-		ret = sof_ipc_tx_message(sdev->ipc,
-					 config->hdr.cmd, config,
-					 config->hdr.size,
-					 &reply, sizeof(reply));
-
-		if (ret < 0) {
-			dev_err(sdev->dev,
-				"error: failed to set dai config for %s\n",
-				dai->name);
+		if (!link_array[config->dai_index]) {
+			ret = sof_ipc_tx_message(sdev->ipc,
+						 config->hdr.cmd,
+						 config,
+						 config->hdr.size,
+						 &reply, sizeof(reply));
+
+			if (ret < 0) {
+				dev_err(sdev->dev,
+					"error: failed to set dai config for %s\n",
+					dai->name);
+
+				devm_kfree(sdev->dev, link_array);
+				return ret;
+			}
 
-			return ret;
+			link_array[config->dai_index] = 1;
```

Comment:
so this is just to keep track of the fact that the DAI was configured already, right?

Reply:
yes, I will add a comment for it